### PR TITLE
[SYCL][Doc] Rename scope-specific variables

### PR DESCRIPTION
**Pull request by Pennycook**
The semantics of work-group-local variables are aligned with C++ thread-local variables but use a different suffix.  This has led to confusion, as some readers assume the "local" in work-group-local refers to the local address space

After considering many alternatives, we settled on the suffix "specific" to describe this concepts: work-group-specific variables are associated with a specific work-group. In future, we expect device-global variables to be renamed to device-specific variables for consistency.

### CODE REVIEW

**Review by artemrad — COMMENTED** [review]
Typo in the filename `sycl_ext_oneapi_worK_group_specific` see capital `K`. 

Also your change does not move or rename the `sycl_ext_oneapi_device_global` spec but rather creates a copy. It is hard to track and review if the changes are properly done, as everything is highlighted as new.

**Comment by Pennycook** [comment]
> Typo in the filename `sycl_ext_oneapi_worK_group_specific` see capital `K`.

Good catch, thanks.  Fixed in https://github.com/intel/llvm/pull/11166/commits/fece8355d3a3b871e9ce83982fd3066ec959cedd.

> Also your change does not move or rename the `sycl_ext_oneapi_device_global` spec but rather creates a copy. It is hard to track and review if the changes are properly done, as everything is highlighted as new.

This is deliberate: we need to separately and simultaneously document what is actually implemented (in experimental) and the proposed name change (in proposed).

I agree GitHub makes this difficult to review, but if you check things out locally you can produce the diff by running:
```sh
git diff --no-index --color-words sycl/doc/extensions/experimental/sycl_ext_oneapi_device_global.asciidoc sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc [hunk 1]
```diff
@@ -0,0 +1,1216 @@
...
+used by permission by Khronos.
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
```
Comment (gmlueck) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc:54]: Let's update this to the suggested text according to the [template](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc#status):

```
This is a proposed update to an existing extension.  Interfaces defined in this
specification may not be implemented yet or may be in a preliminary state.  The
specification itself may also change in incompatible ways before it is
finalized.  *Shipping software products should not rely on APIs defined in this
specification.*  See
link:../supported/sycl_ext_oneapi_myextension.asciidoc[here] for the existing
extension, which is implemented.
```

**Review by gmlueck — COMMENTED** [review]
I like the new naming style.

What are your thoughts about implementation?

1. Should we implement the new name as "experimental", replacing the old name?
2. Should we implement the new name as "experimental", keeping the old name as deprecated?
3. Should we rename the type and also move the extension to "supported" at the same time, allowing us to keep the experimental version with the old name?

If we think we're ready to move this to "supported" anyways, doing it all together seems like less hassle for users.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc (hunk 1 above)
Comment (Pennycook) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc]: Fixed in ec7097a6dd34c0d9b6401f94b197f5da80d7b829.

**Comment by Pennycook** [comment]
> If we think we're ready to move this to "supported" anyways, doing it all together seems like less hassle for users.

I agree, we should do 3, to ensure that developers only go through the pain of renaming once.

I'm not sure on timeline: I haven't reviewed `device_global`'s interface closely (only the name).  It's probably worth going through to check if `device_specific` and `work_group_specific` have any significant differences in their API, and whether that's something we would want to address before moving to "supported".  If there _are_ differences in the APIs, I expect we'd have to justify that (or adjust the APIs) when proposing these features as a KHR.

**Review by gmlueck — APPROVED** [review]

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc [hunk 2]
```diff
@@ -0,0 +1,1217 @@
...
+
+=== Properties for device-specific variables
+
+The `device_specific` class supports several compile-time-constant properties.
+If specified, these properties are included in the `PropertyListT` template
+parameter as shown in this example:
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi::experimental;
```
Comment (artemrad) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc:509]: The `device_specific` spec relies on the `properties` spec, which is in the experimental namespace. Shouldn't we move the `properties` spec, first to the supported namespace before moving the `device_specific` spec?

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc [hunk 3]
```diff
@@ -0,0 +1,1217 @@
...
+----
+init_mode
+----
+a|
+This property is only meaningful when used with an FPGA device.  It is ignored
+for other devices.  The following values are supported:
+
+* `reprogram`: Initialization is performed by reprogramming the device.  This
+  may require more frequent reprogramming but may reduce area.
+* `reset`: Initialization is performed by sending a reset signal to the device.
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc:688]: If `init_mode::reset` is still not implemented by the time we want to move this extension to "supported", we could remove the `init_mode` property from the specification.  We can add it back later in version 2 of the extension.

**Comment by Pennycook** [comment]
@artemrad: Are the latest changes to device-global done, now?  Could you please help to update the device-specific specification to match any recent changes?

**Comment by artemrad** [comment]
@Pennycook I know of two more issues that might impact this. I want your input on how you want to deal with this. 
1. https://github.com/intel/llvm/pull/11107 This case tracks the implementation of the extension that the spec talks about - supporting consteval constructors.
2. There seems to be internal disagreement on the FPGA specific property `implement_in_csr`. This might get renamed in the future, once quorum is reached. 

My proposal is that we break up the `experimental` spec into three parts.

 First part is the majority of the spec, but without the FPGA specific properties. Specifically it will only contain  `device_image_scope` and `host_access` properties. This spec will be in in the `supported` name space. This spec will not have #1 above documented, as it's implementation is close to being done, but it has not been used much, to give us confidence that we won't change the semantics. 

Part two, will be an extension in the `experimental` folder that talks about the addition of consteval constructors. It will reference the spec in part 1. 

Part three will be an extension in the `experimental` folder that talks about the FPGA specific properties applied to device_global, specifically `init_mode` and `implement_in_csr`. These have yet to be fully implemented by our team, and may change if major issue come up. This spec will reference the spec in part 1.

What do you think? 
@GarveyJoe

**Comment by Pennycook** [comment]
Thanks, @artemrad.

Splitting the specification into separate parts sounds good to me.  I think the only thing I feel strongly about here is that the name change should happen before it becomes "supported".

Perhaps the right thing to do is to have somebody from the FPGA side update this PR, and turn sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc into something reflective of your "first part".  We can then work on implementing that proposed spec, and move it to "supported" when it's done.

**Comment by againull** [comment]
@Pennycook As far as I understood additional changes will be needed in scope of this PR. That's why turning this into a draft for now so that it doesn't get merged. Very sorry if my understanding is wrong, let me know if so. Please mark as ready for review when it will be ready for merge.

**Comment by Pennycook** [comment]
@againull: I've removed the file that was causing problems. I'd like to get the work-group-specific changes merged in before anybody implements the extension.  We'll need to revisit the renaming of device-global when the API is stable.

FYI: @artemrad, @GarveyJoe

**Pull request merged**